### PR TITLE
check type visibility for quickfixes

### LINKED ISSUE #14105: Implement missing members sometimes writes out inaccessible types
```ts
function foo() {
  abstract class C {
    abstract myMethod(): C;
  }

  return C;
}

class D extends foo() {
}
```

Right now, we enable a codefix for implementing missing members of the class `C`. However, `C` is inaccessible, and the generated code will be something like the following

```ts
function foo() {
  abstract class C {
    abstract myMethod(): C;
  }

  return C;
}

class D extends foo() {
  myMethod(): C {
    throw new Error('Method not implemented.');
  }

}
```

which is invalid.

I'd argue that the code fix shouldn't be made available.

### PR DESCRIPTION
**Pull request by aozgaa**
Fixes #14105.

Feedback on type de-structuring over types here will be used in approaching TypeNode synthesis.

### CODE REVIEW

**Comment by aozgaa** [comment]
Can you please review @vladima, @mhegazy , @billti, @zhengbli ?

File: tests/cases/fourslash/codeFixClassExtendAbstractAccessibilityTuple.ts [hunk 1]
```diff
@@ -0,0 +1,18 @@
+/// <reference path='fourslash.ts' />
+
+//// namespace N {
+////     const enum Inaccessible {
+////         member
+////     };
+////     export abstract class A {
+////         // abstract x: [number, Inaccessible, string];
+////         // abstract y: [number, Inaccessible.member];
```
Comment (billti) [inline tests/cases/fourslash/codeFixClassExtendAbstractAccessibilityTuple.ts]: Why are the above 2 lines commented out?

**Comment by billti** [comment]
The tests cover a range of patterns at one level, but what about multiple levels (e.g. namespaces within namespaces), or types that are imported (e.g. file1 has `import {typeA} from 'foo'` and exports a class C that uses typeA, does file2 that imports class C generate the correct quickfix?

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -1985,6 +1986,100 @@ namespace ts {
...
+
+            return isTypeAccessibleWorker(type, /*inObjectLiteral*/ false, /*inTypeAlias*/true);
+
+            function isTypeAccessibleWorker(type: Type, inObjectLiteral: boolean, inTypeAlias: boolean): boolean {
+                if (!type) {
+                    return false;
+                }
+
+                if (inTypeAlias && type.aliasSymbol) {
+                    return isSymbolAccessible(type.aliasSymbol, enclosingDeclaration, SymbolFlags.Type, /*shouldComputeAliasesToMakeVisible*/false).accessibility === SymbolAccessibility.Accessible
```
Comment (billti) [inline src/compiler/checker.ts]: You use `SymbolFlags.Type` here, and `meaning` below (which you created as a const for SymbolFlags.Type on line 1990). Either get rid of the const, or use it consistently.

File: tests/cases/fourslash/codeFixClassExtendAbstractAccessibilityTuple.ts (hunk 1 above)
Comment (aozgaa) [inline tests/cases/fourslash/codeFixClassExtendAbstractAccessibilityTuple.ts]: Oversight from some edits part way through. Fixed. Thanks!

File: src/compiler/checker.ts (hunk 2 above)
Comment (aozgaa) [inline src/compiler/checker.ts]: Removed `meaning`.

**Comment by aozgaa** [comment]
Added tests per feedback.

File: src/compiler/checker.ts [hunk 3]
```diff
@@ -133,9 +133,18 @@ namespace ts {
             signatureToString: (signature, enclosingDeclaration?, flags?, kind?) => {
                 return signatureToString(signature, getParseTreeNode(enclosingDeclaration), flags, kind);
             },
+            signatureToAccessibleString: (signature, enclosingDeclaration?, flags?, kind?) => {
```
Comment (mhegazy) [inline src/compiler/checker.ts]: why do you need these on the checker? `getSymbolDisplayBuilder` is already visible on the `TypeChecker` interface. I would just move these to codefixs/helpers.

**Comment by aozgaa** [comment]
Once we get `createTypeNode` in, the fixes will be converted to work with the normal emit pipeline rather than the `xToString` methods and their writers, so we will need to come up with a different solution then.

**Pull request closed without merging**